### PR TITLE
ARROW-483: [C++/Python] Provide access to "custom_metadata" Field attribute in IPC setting

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: python/pyarrow/tests/test_parquet.py [hunk 1]
```diff
@@ -425,6 +426,11 @@ def test_read_single_row_group():
     result = pa.concat_tables(row_groups)
     tm.assert_frame_equal(df, result.to_pandas())
 
+
+@parquet
+def test_read_single_row_group_with_column_subset():
```
Comment (cpcloud) [inline python/pyarrow/tests/test_parquet.py]: This test will fail until we decide how to handle column subsetting.

**Comment by jreback** [comment]
it *might* make sense *not* to save a regular ``RangeIndex`` Index directly (e.g. 0...len), and just indicate this in the meta-data.

**Comment by cpcloud** [comment]
I think it will ease merging and make things a little less churny if I split this patch into one that implements the metadata API but not the index machinery and one that uses the metadata to reconstruct the index.

**Comment by cpcloud** [comment]
I'll open a new JIRA for the index reconstruction

**Pull request closed without merging**